### ISSUE TITLE
Determine how to handle deprovision requests to an instance with bindings

### ISSUE THREAD

**Issue by pmorie** [kubernetes-incubator/service-catalog#820]
This issue is to determine how to handle deprovision requests to instances with bindings.  The primary discussion initially on this issue was in F2F meetings.  There were two primary schools of thought:

- Delete all bindings
- Disable new bindings to the instance, automatically delete instance when existing bindings were deleted.

**Comment by arschles** [comment]
See https://github.com/kubernetes-incubator/service-catalog/issues/801#issuecomment-299994481 for a description of what we decided in our last face-to-face meeting. I am leaving this open for now, as this issue represents a specific subset of #801. Both may not be closed at the same time, however.

cc/ @pmorie @duglin

**Comment by arschles** [comment]
Here is a very brief (since I only have 5 minutes before our next design meeting) proposal for allowing users to specify the deletion behavior of `Instance`s at creation time. It may be combined with other methods if we deem it appropriate. If there is interest in this method, I will write a more detailed proposal.

We can add a field to the `Instance` spec called `deleteBehavior`, which has two possible behaviors: `forceCascade` or `failOnBindings`. In the former case, deleting an instance will force-delete all `Binding`s that reference the instance, without confirming with the person who deleted it. In the latter, the deletion simply will fail if there are bindings that reference the instance. This field will be optional, and default to `failOnBindings` if it is missing.

**Comment by ash2k** [comment]
Is this behaviour specified in the OSB spec? i.e. what happens when marketplace tells the provider to deprovision an instance which has bindings? Should it be an error?

**Comment by arschles** [comment]
I don't believe it is; the [deprovisioning section](https://github.com/openservicebrokerapi/servicebroker/blob/master/spec.md#deprovisioning) does not have any language indicating what happens.

**Comment by arschles** [comment]
Here's what we decided in the design meeting:

- Set deletion timestamp on the instance
- Add a condition in the instance status indicating which bindings need to be deleted before the instance will be deleted
- Block creation of new bindings against the instance (in an admission controller)


We are also going to bring up the concept of [binding updates](https://github.com/openservicebrokerapi/servicebroker/issues/123) in the OSB API Working Group. We will want to block updates as well.

**Comment by jboyd01** [comment]
I'm investigating, will update with details.

**Comment by pmorie** [comment]
I think we need to go a little further on the design for this.  [This comment](https://github.com/kubernetes-incubator/service-catalog/issues/820#issuecomment-317554625) is a fine start but I see a couple remaining gaps:

- How does the controller differentiate between an instance it can finalize vs instance with bindings?  Is it a new condition?  Is it a new finalizer?
- However we manage representing to the controller that it _can't_ finalize an instance yet, how does that information get managed?  If I delete an instance with 3 bindings, how does the instance get updated so that the controller know's it's okay to deprovision once the last binding is deleted?

Here's another entanglement.  This sounds very much to me like garbage collection, or at least seems like it's superficially related.  We might be able to leverage something in the machinery to help us with this.

**Comment by ironcladlou** [comment]
I'm coming to this with only scraps of knowledge about service catalog, but I've been deep into the Kube garbage collector recently and have had some brief discussions with @jboyd01.

If some bindings require a manual user intervention which should block the deletion of the instance during a cascading delete, perhaps finalizers could be employed to allow the Kube garbage collector to do the heavy lifting.

For any binding which requires manual intervention to allow deletion, add a finalizer (e.g. `UserDeletionApproval`) to the binding. Now, when the instance is deleted with the [foreground deletion policy](https://kubernetes.io/docs/concepts/workloads/controllers/garbage-collection/#foreground-cascading-deletion), the instance will remain until all bindings are first deleted. Those bindings without the `UserDeletionApproval` finalizer will be deleted automatically by the garbage collector. Those with the finalizer will linger until some other process removes the finalizer, at which point binding will be deleted. Once all bindings are gone, the instance will be deleted by the garbage collector.

One possible hiccup here would be ensuring the deletion grace period is sufficiently long to allow the instance (and the dependents) to remain in a deleting state long enough to support the approval use case.

This could be prototyped easily and immediately with a Kube 1.8 build.

**Comment by ironcladlou** [comment]
Regarding admission control to block new dependent bindings of an instance currently being deleted, I would suggest the possibility of creating bindings in an initially benign state which are never acted upon until they are "accepted". They would never be accepted if their owner is being deleted, and would eventually get GC'd. If you already have an external initializer/admission controller then maybe it makes sense to go ahead and use it, but I'd be wary of introducing one just for this purpose.

**Comment by pmorie** [comment]
The desired behavior is for the user to manually delete the bindings.

**Comment by pmorie** [comment]
@ironcladlou one other wrinkle here is that we are currently on the 1.7 machinery

**Comment by ironcladlou** [comment]
@pmorie 
>The desired behavior is for the user to manually delete the bindings.

From a user's perspective, is it important the user be able to actually issue a DELETE on the binding? Or is it acceptable for them to "approve" a delete via some other operation which causes the binding to be deleted indirectly?

**Comment by ironcladlou** [comment]
For example, can the user "delete" the binding via an update to the binding which removes the finalizer?

**Comment by ironcladlou** [comment]
@pmorie 
>one other wrinkle here is that we are currently on the 1.7 machinery

Unfortunately, it seems unlikely GC support for custom resources will be backported to 1.7

**Comment by pmorie** [comment]
@ironcladlou the user DELETEs the binding to cause unbinding.  Bindings rely on graceful deletion to work correctly in the OSB API.

**Comment by jboyd01** [comment]
# Design Proposal

## Abstract

If a service instance with existing bindings is deleted it would break any applications that are using the binding.  This document addresses how Service Catalog will behave when a user deletes a service instance that has bindings and it discusses the possible life cycle stages for the instance and its bindings.

## Proposed Behaviors

### Long Term
When a user attempts to delete an instance (ie **kubectl delete instance**) Catalog will check if any bindings are associated with the instance.  If there are, the operation will fail with an error indicating that the instance is in use.  The user can repeat the operation with a **-f**orce flag.  In this case, the instance will be marked as soft deleted.  Instances that don't have bindings will be deleted immediate with or without the force flag.

### Short Term
For the immediate term, we do not fail the delete if there are bindings.  We do not implement a force argument for  the delete instance.  In the reconcile operation, if there are bindings, we set the ready condition to false and add a message indicating the delete in in progress but it's waiting for existing bindings to be deleted fist.  Eventually, after all bindings have been deleted by operator commands, the periodic execution of reconcile will finish deleting the instance.

### Common to Short and Long term
Applications with bindings to soft deleted instances will continue to work.  Attempts to create new bindings or modify existing bindings for the soft deleted instance will be blocked.  When all bindings for the soft deleted instance have been deleted, the instance will be fully deleted.


## High Level Design

1.  Service Catalog accepts valid delete binding commands and sets the delete timestamp on the specified instance (no change here).

2.  Attempts to create new bindings associated with a soft deleted instance will result in an error as will attempting to update an existing binding associated with the soft deleted instance. (new admission controller)

3. The instance reconciliation logic will be updated to check for existing bindings and if found, set the ready condition to false and add a message indicting the bindings block the delete.

As a follow on to item 3, a mechanism needs to be in place to delete soft deleted instances once all associated bindings are removed.  There are several options here:

-A.  Instance conciliation realizes there are no more bindings and completes the delete operation

- B.  Make the binding delete operation responsible for instance cleanup
    1.  When a binding is deleted, if the instance is marked for soft deletion, check if there are any other bindings associated with instance.  If there are none, delete the instance.

- C.  Utilize Garbage Collection
    1.  Use [foreground cascading deletion](https://kubernetes.io/docs/concepts/workloads/controllers/garbage-collection/#foreground-cascading-deletion).  Bindings would be marked as blocking dependents (parent can't be deleted prior to the blocking dependents).  We would use UserDeletionApproval finalizers on the associated bindings that would prevent GC from actually deleting any bindings, but once the bindings are deleted (by user delete actions) the soft deleted instance would be removed by GC.
    2.  One possible hiccup here would be ensuring the deletion grace period is sufficiently long to allow the instance (and the dependents) to remain in a deleting state long enough to support the approval use case which could be indefinite.
    3.  GC for custom resources is first available in 1.8, it would need to be backported to 1.7 which seems [unlikely](https://github.com/kubernetes-incubator/service-catalog/issues/820#issuecomment-323160391)

Option A is chosen for short term as it is simple.  Option B would result in a more immediate deletion of the instance, but it involves more work and spreads the instance life cycle logic into the binding.  Option C (GC) has numerous risks including a dependency to get functionality back ported to 1.7.  Perhaps we'll revisit GC in the future.


We had previous discussion about [allowing users to specify the deletion behavior of Instances at creation time, ie forceCascade or failOnBindings](https://github.com/kubernetes-incubator/service-catalog/issues/820#issuecomment-316179837) and [adding a condition in the instance status indicating which bindings need to be deleted before the instance will be deleted](https://github.com/kubernetes-incubator/service-catalog/issues/820#issuecomment-317554625).  These alternate approaches will not be utilized.

## Design Details
- create an **admission controller** to block adding bindings to a soft deleted instance
- update the instance reconciliation:
    - if the instance is marked as soft deleted, check if there are any other bindings associated with the instance.  If there are, set the Ready condition to false and set the condition message to indicate the instance deletion is blocked by the presence of related bindings.  If there are no related bindings, complete the deletion of the instance


From initial proof of concept, this works, results in the following Instance condition & msg after a blocked deletion on the instance:
```yaml
    conditions:
    - lastTransitionTime: 2017-08-24T15:01:02Z
      message: Deprovision call failed. Deprovisioning Instance test-ns/ups-instance
        blocked by existing binding test-ns/ups-instance
      reason: DeprovisionCallFailed
      status: "False"
      type: Ready
```

**Comment by ironcladlou** [comment]
I haven't had a chance to read the new design proposal yet, but getting the notification reminded me that we need to be careful now to reconsider any design which uses conditions, which are proposed to [be eliminated](https://github.com/kubernetes/kubernetes/issues/7856#issuecomment-324219862) as a concept.

**Comment by jboyd01** [comment]
Thanks @ironcladlou, great to know.  There was early discussion in this item about using new conditions to track the associated bindings, but I've indicated in the design proposal that that it is out of scope.

**Comment by pmorie** [comment]
Closed by #1159
